src/lib.rs

```diff
@@ -1408,6 +1538,356 @@ where
+    client_id: &'a ClientId,
+    client_secret: Option<&'a ClientSecret>,
+    extra_params: Vec<(Cow<'a, str>, Cow<'a, str>)>,
+    token_url: Option<&'a TokenUrl>,
+    dev_auth_resp: &'a DeviceAuthorizationResponse<EF>,
+    time_fn: Arc<dyn Fn() -> DateTime<Utc> + 'a + Send + Sync>,
+    _phantom: PhantomData<(TR, TT, EF)>,
+}
+
+impl<'a, TR, TT, EF> DeviceAccessTokenRequest<'a, TR, TT, EF>
+where
+    TR: TokenResponse<TT>,
+    TT: TokenType,
+    EF: ExtraDeviceAuthorizationFields,
+{
+    ///
+    /// Appends an extra param to the token request.
+    ///
+    /// This method allows extensions to be used without direct support from
+    /// this crate. If `name` conflicts with a parameter managed by this crate, the
+    /// behavior is undefined. In particular, do not set parameters defined by
+    /// [RFC 6749](https://tools.ietf.org/html/rfc6749) or
+    /// [RFC 7636](https://tools.ietf.org/html/rfc7636).
+    ///
+    /// # Security Warning
+    ///
+    /// Callers should follow the security recommendations for any OAuth2 extensions used with
+    /// this function, which are beyond the scope of
+    /// [RFC 6749](https://tools.ietf.org/html/rfc6749).
+    ///
+    pub fn add_extra_param<N, V>(mut self, name: N, value: V) -> Self
+    where
+        N: Into<Cow<'a, str>>,
+        V: Into<Cow<'a, str>>,
+    {
+        self.extra_params.push((name.into(), value.into()));
+        self
+    }
+
+    ///
+    /// Specifies a function for returning the current time.
+    ///
+    /// This function is used while polling the authorization server.
+    ///
+    pub fn set_time_fn<T>(mut self, time_fn: T) -> Self
+    where
+        T: Fn() -> DateTime<Utc> + 'a + Send + Sync,
+    {
+        self.time_fn = Arc::new(time_fn);
+        self
+    }
+
+    ///
+    /// Synchronously polls the authorization server for a response, waiting
+    /// using a user defined sleep function.
+    ///
+    pub fn request<F, S, RE>(
+        self,
+        http_client: F,
+        sleep_fn: S,
+        timeout: Option<Duration>,
+    ) -> Result<TR, RequestTokenError<RE, DeviceCodeErrorResponse>>
+    where
+        F: Fn(HttpRequest) -> Result<HttpResponse, RE>,
+        S: Fn(Duration) -> (),
+        RE: Error + 'static,
+    {
+        let details = self.dev_auth_resp;
+
+        // Calculate the request timeout - if the user specified a timeout,
+        // use that, otherwise use the value given by the device authorization
+        // response.
+        let timeout_dur = timeout.unwrap_or(details.expires_in());
+        let chrono_timeout = chrono::Duration::from_std(timeout_dur)
+            .map_err(|_| RequestTokenError::Other("Failed to convert duration".to_string()))?;
+
+        // Calculate the DateTime at which the request times out.
+        let timeout_dt = (*self.time_fn)().checked_add_signed(chrono_timeout).ok_or(
+            RequestTokenError::Other("Failed to calculate timeout".to_string()),
+        )?;
```

Comment:
it's only ~10 lines of duplication, but given the complexity here, mind adding a `self.compute_timeout(timeout.as_ref())` (or similar) function?

Reply:
Done.